refactor: Remove all old Temporal loggers

tomasfarias opened the pull request:
## Problem

Our logging setup for Temporal is currently divided into two "logging pipelines". One of these pipelines is referred to as "sync loggers" or "sync logging" as it consists of starting a synchronous Kafka producer to produce logs to ClickHouse. This duplicates pretty much all the work that the "async" or "regular" logging pipeline does, while also introducing some issues:

For starters, the synchronous Kafka producer is started on every attempt to get a logger, and only when structlog is not configured.

This is problematic:
* `structlog.is_configured` may return `True` if Django configured it first, thus no producer is ever started.
  * Luckily, we seem to be winning the race, but who knows, we may be losing it sometimes.
* If `structlog.configure` is not threadsafe, then that means we could be inadvertently starting multiple producers: As we check for `is_configured` is `False` if another thread is already configuring it, but not locking, we could be starting the configuration again.
* Why do we need to check this on every attempt to get a logger? This just adds unnecessary overhead. Logging should be configured just once at the start of the worker and then get out of the way.

Moreover, the "sync loggers" were never required in the first place:
* The regular logging pipeline runs an asynchronous Kafka producer in the worker/activity event loop. Even if the activities from where we log are synchronous and running in another thread, you can still pass data to the producer's queue using something like `asyncio.run_coroutine_threadsafe(queue.put(msg), loop)` (this doesn't block the activity thread, it returns immediately!).

There is also an anti-pattern I keep seeing, which is passing loggers as arguments or initializing loggers more than once:
* This is done because a lot of the times we are not binding context variables.
* That's why the `structlog.contextvars.bind_contextvars` function exists: Once something is bound with it, ALL loggers get the variables added to their context. Yes, this context is copied over to new asyncio tasks AND threads, no need to pass loggers around.

Moreover, there is a million functions:
* `configure_temporal_worker_logger`, `bind_temporal_org_worker_logger`, `bind_temporal_worker_logger_sync`, `bind_temporal_worker_logger`, `get_internal_logger`.
  * New and old engineers working with Temporal have a hard time picking the right one.
  * The concept of "external" and "internal" logs is only valid in certain cases (when the logs are actually read by external users).

Finally, even the async pipeline is not without issues:
* It forces logs going to clickhouse to come from a logger named "EXTERNAL". This means you can't query by logger in other tools if you use the same logger for everything.

<!-- Who are we building for, what are their needs, why is this important? -->

<!-- Does this fix an issue? Uncomment the line below with the issue ID to automatically close it when merged -->
<!-- Closes #ISSUE_ID -->

## Changes

This PR is here to refactor the entire logging pipeline:
* I have removed the sync kafka producer, sync queue, sync configuration functions.
  * The async task is started once when initializing workers, and that's it.
  * Since the pipeline is configured always when a worker starts, the `TEMPORAL_USE_EXTERNAL_SETTING` setting is removed as its no longer necessary.

* I have removed all the functions related to obtaining a logger. The way to obtain a logger is now `structlog.get_logger`.
  * For convenience, `get_produce_only_logger`, `get_write_only_logger`, and `get_logger` functions exist in `temporal.common.logger` but these contain just a single call to `structlog.get_logger`, nothing else. Their only purpose is to provide better type hints and docstrings.

* All loggers can now write to stdout and/or produce logs to clickhouse.
  * Instead of "external" and "internal" loggers we now define loggers by what they do: Either write (to a file) or produce (to kafka/clickhouse). This is independent of what is ultimately done with the logs and should better reflect the consequence of using one log or the other.
  * By default, a logger you obtain with `structlog.get_logger` will do both.
  * `get_produce_only_logger` and `get_write_only_logger` can be used to obtain loggers that only do one of the two things.
  * Using a write-only logger offers slightly better performance, as we save the production part of the processing. But this is very minimal.
  * With the default logger, we can also override the behavior on a log by log basis, by passing the `produce_only=True` or `write_only=True` arguments.

* It is no longer required to call `bind_contextvars` to populate the context with temporal variables. These are populated automatically when available.

* Loggers now work from workflow context, although producing logs from a workflow is not supported as the Temporal workflow tasks run on a different event loop that cannot run threads or do any I/O.

* I have replaced combed through all modules and replaced all calls to the old functions for calls to the new functions.
* For some reason, we were importing `FilteringBoundLogger` from a module where it's not defined. I've changed these imports to use `structlog.types`.

## Todo

* I am not fixing all the instances where loggers are passed as arguments. There are too many and I lack some context of what needs to be bound to what context.

* README updates are pending.

<!-- If there are frontend changes, please include screenshots. -->
<!-- If a reference design was involved, include a link to the relevant Figma frame! -->

## How did you test this code?

I added unit tests for all steps in the logging pipeline.
<!-- Briefly describe the steps you took. -->
<!-- Include automated tests if possible, otherwise describe the manual testing routine. -->

<!-- Docs reminder: If this change requires updated docs, please do that! Engineers are the primary people responsible for their documentation. 🙌 -->

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

tomasfarias committed:
refactor: Remove all old loggers

tomasfarias committed:
fix: Use correct fixture name

tomasfarias committed:
fix: Remove all remaining references to TEMPORAL_USE_EXTERNAL_LOGGER

tomasfarias committed:
fix: Let the loop run

tomasfarias committed:
fix: Add missing await

tomasfarias committed:
fix: Add a type hint

tomasfarias committed:
refactor: Rename external logger to log_entries logger

tomasfarias committed:
feat: Make logger tests generic for async and sync functions

tomasfarias committed:
feat: New logging pipeline with tests

tomasfarias committed:
refactor: Everything to use new loggers

tomasfarias committed:
fix: Add missing import

tomasfarias committed:
fix: Annoying type fixes

tomasfarias committed:
fix: Add missing logging methods

tomasfarias committed:
fix: Configure logger on all temporal tests

tomasfarias committed:
fix: Do not produce logs in tests

tomasfarias committed:
fix: Use correct key in test

tomasfarias commented on posthog/temporal/tests/common/test_logger.py:
```diff
@@ -0,0 +1,794 @@
+        activity_type="test-activity",
+        attempt=random.randint(1, 10000),
+        task_queue="batch-exports-task-queue",
+        workflow_id=f"{ID}-Backfill-{dt.datetime.now(dt.UTC)}",
+        workflow_namespace="prod-us",
+        workflow_type="backfill-batch-export",
+        workflow_run_id=str(uuid.uuid4()),
+    ),
+    # Data warehouse
+    ActivityInfo(
+        activity_id=random.randint(1, 10000),
+        activity_type="test-activity",
+        attempt=random.randint(1, 10000),
+        task_queue="data-warehouse-task-queue",
+        workflow_id=f"{ID}-{dt.datetime.now(dt.UTC)}",
+        workflow_namespace="prod-us",
+        workflow_type="external-data-job",
+        workflow_run_id=str(uuid.uuid4()),
+    ),
+    ActivityInfo(
+        activity_id=random.randint(1, 10000),
+        activity_type="test-activity",
+        attempt=random.randint(1, 10000),
+        task_queue="data-warehouse-task-queue",
+        workflow_id=f"{ID}-compaction",
+        workflow_namespace="prod-us",
+        workflow_type="deltalake-compaction-job",
+        workflow_run_id=str(uuid.uuid4()),
+    ),
+    ActivityInfo(
+        activity_id=random.randint(1, 10000),
+        activity_type="test-activity",
+        attempt=random.randint(1, 10000),
+        task_queue="data-warehouse-task-queue",
+        workflow_id=f"{ID}-{dt.datetime.now(dt.UTC)}",
+        workflow_namespace="prod-us",
+        workflow_type="data-modeling-run",
+        workflow_run_id=str(uuid.uuid4()),
+    ),
+]
```
doesn't have anything to do with ui snapshots.

tomasfarias commented on posthog/temporal/tests/common/test_logger.py:
```diff
@@ -0,0 +1,794 @@
+    def log_sync():
+        """A simple function that logs with multiple loggers."""
+        logger = structlog.get_logger("test_logger_produces_to_log_queue")
+        produce_only = structlog.get_logger("test_produce_only", False, True)
+        write_only = structlog.get_logger("test_write_only", True, False)
```
I can't. `structlog.get_logger` treats keyword arguments as initial values, not configuration parameters.

This is why convenience functions exist. But this is a test specific to the logger module.

This is just noise from a pretty useless tool.

tomasfarias committed:
feat: Update README and docstrings

tomasfarias commented on products/batch_exports/backend/temporal/destinations/http_batch_export.py:
```diff
@@ -121,15 +121,15 @@ async def maybe_resume_from_heartbeat(inputs: HttpInsertInputs) -> str | None:
         # This is the error we expect when there are no activity details as the sequence will be
         # empty.
-        logger.debug(
+        LOGGER.debug(
             "Did not receive details from previous activity Excecution. Export will start from the beginning %s",
             interval_start,
         )
```
Huh, no? The `bind_contextvars` is called before this function is called. The context will be populated by the time we get here.

Gilbert09 commented on posthog/temporal/data_imports/pipelines/pipeline_sync.py:
```diff
@@ -79,8 +82,8 @@ def validate_schema_and_update_table_sync(
     """
-
-    logger = bind_temporal_worker_logger_sync(team_id=team_id)
+    bind_contextvars(team_id=team_id)
+    logger = LOGGER.bind()
```
Nit: it would be nice to be able to do this in a single function call - `logger = LOGGER.bind(team_id=team_id)`

tomasfarias replied:
It's not semantically equivalent: `bind_contextvars` binds `team_id` to the local context, where as `LOGGER.bind` only binds it to the logger itself.

In the first case, any logger in the same local context (even across threads and asyncio tasks) will have `team_id` bound.

In the second case, the bind only applies to the one logger object. 

Anyways, all that being said, I checked and the function we are replacing here `bind_temporal_worker_logger_sync` did something closer to `LOGGER.bind`. So, I'll swap this to use `LOGGER.bind` just to stick closer to the old behavior.

Gilbert09 commented on posthog/temporal/data_modeling/run_workflow.py:
```diff
@@ -997,8 +1002,9 @@ def __init__(self, invalid_input: str):
+    logger = LOGGER.bind()
 
-    logger = await bind_temporal_worker_logger(inputs.team_id)
     await logger.adebug(f"starting build_dag_activity. selectors = {[select.label for select in inputs.select]}")
```
From the readme, it sounds like it's now better to do synchronous log calls, is that right? (I can change all of these after release if so)

tomasfarias replied:
In this PR I didn't want to change anything in the current behavior though. So I've kept all the async usage.

Gilbert09 commented on posthog/temporal/common/logger.py:
```diff
@@ -1,277 +1,542 @@
-    """Bind any variables to the context, including base Temporal variables."""
-    temporal_context = get_temporal_context()
-    structlog.contextvars.bind_contextvars(**temporal_context, **kwargs)
+    log = debug = info = warn = warning = process  # type: ignore
+    fatal = failure = err = error = critical = exception = process  # type: ignore
```
Is there any more context to this? 👀

Gilbert09 reviewed (approved):
Look great - good work!

tomasfarias committed:
refactor: Stick closer to old behavior when binding logger

tomasfarias committed:
refactor: Remove inheritance from structlog.WriteLogger

Previous implementation was breaking Liskov substitution principle as
NamedLogger subclasses could not be used in place of
strucltog.WriteLogger.

So, this commit prefers composition over inheritance by holding a
WriteLogger as an attribute of Logger. Write calls are then dispached
to the underlying logger.

This removes the need for a NamedLogger class entirely, and removes
some mypy comments.

tomasfarias committed:
fix: Import paths use new functions

tomasfarias committed:
fix: README use correct parameter
